Python/pythonrun.c

```diff
@@ -1164,6 +1164,36 @@ print_exception_notes(struct exception_print_context *ctx, PyObject *notes)
     return -1;
 }
 
+static int
+get_exception_notes(struct exception_print_context *ctx, PyObject *value, PyObject **notes) {
+    PyObject *note = NULL;
+
+    if (_PyObject_LookupAttr(value, &_Py_ID(__notes__), notes) < 0) {
+        PyObject *type, *errvalue, *tback;
+        PyErr_Fetch(&type, &errvalue, &tback);
```

Comment:
I think actually we need to normalise the exception here. 

Try adding a test where ``__getattr__`` raises by type: ``raise ValueError``.

Reply:
I'm not sure what we are supposed to see by using `raise ValueError`, I tried to change the current test to use like this:

```py
        class BrokenException(Exception):
            broken = False
            def __getattr__(self, name):
                if self.broken:
                    raise ValueError                 # <= changed raise style
                raise AttributeError(name)

        e = BrokenException(123)
        vanilla = self.get_report(e)
        e.broken = True
        self.assertEqual(
            self.get_report(e),
            vanilla + "Ignored error getting __notes__: ValueError()\n")   # <= changed message
```

and the test passed.

The doc of [PyErr_NormalizeException](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_NormalizeException) mentions:

> Under certain circumstances, the values returned by [PyErr_Fetch()](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_Fetch) below can be “unnormalized”, meaning that *exc is a class object but *val is not an instance of the same class. This function can be used to instantiate the class in that case

does this mean that `errvalue` is always the expected, but `type` might be a different class ? Here we are only using `errvalue` and not `type`, that might be a reason. I don't really understand this and I'm guessing a lot here :)

In any case, if we add `PyErr_NormalizeException(&type, &errvalue, &tback);` here, everything seems also OK, so if you think we should add it, let's add it. Next question would be "do we also want to do it for 3.11 ?"

Reply:
The type is always set, the value may not be. I think we should normalise (in 3.11 too) but I'm surprised the test passes, TBH.

Reply:
Thank you, I'm a bit surprised as well, but it seems better to call `PyErr_NormalizeException`. I pushed a commit here and made another PR to have these changes in 3.11 https://github.com/python/cpython/pull/114379

Reply:
It seems the exception is being normalised in `set_attribute_error_context`, this is why the test passes. But it's goto practice to normalise whenever you call `PyErr_Fetch`, so let's do it.